Feature/excel validation framework

AdrienHorgnies opened the pull request:
standard validation

Very similar to what you already saw except that I used a more recent version of poijoi and tweaked the logic a bit:
* I used a Row < Sheet < Workbook hierarchy without any further model definition classes as I deemed those three sufficient in that regards. The only place I feel a bit of regret for is where I feed the content of the workbook to the workbook that served the purpose of model.
* I extracted some code to methods and made some improvement there and there.

Commit:
implement generic classes for spreadsheets

Commit:
add service to check validate workbook

Commit:
dummy endpoint to test framework

Commit:
example endpoint and some standard validation

Commit:
create issues to resolve the todo items

Commit:
travis configuration file

Commit:
cache maven repository

Commit:
disable eureka during test to avoid log pollution, closes #8

rlaloux commented on src/main/java/com/altissia/clickandrun/domain/spreadsheet/validation/FieldValidation.java:
```diff
@@ -0,0 +1,84 @@
+	 */
+	@Override
+	public int compareTo(FieldValidation other) {
+		int compare = this.getField().compareTo(other.getField());
```
What about `null`?

You might want to use a null-aware comparator such as:

```java
private static Comparator<String> comparator = Comparator.nullsFirst(String::compareTo);
```

AdrienHorgnies replied:
done

rlaloux commented on src/main/java/com/altissia/clickandrun/domain/spreadsheet/Workbook.java:
```diff
@@ -0,0 +1,83 @@
+ */
+public abstract class Workbook {
+
+    protected boolean ignoringSuperfluousHeaders;
```
Simplify this property name. `ignoreUnknownHeaders` would be more suitable.

AdrienHorgnies replied:
You sure ? Because the standard getter method name for a boolean is the property prefixed with "is" I prefer "isIgnoringSuperfluousHeaders" to "isIgnoreSuperfluousHeaders".

rlaloux commented on src/main/java/com/altissia/clickandrun/domain/spreadsheet/concrete/registration/RegistrantRow.java:
```diff
@@ -0,0 +1,75 @@
+import javax.validation.constraints.Pattern;
+import java.util.Objects;
+
+public class RegistrantRow extends Row {
```
Why "RegistrantRow" and not "RegistrationRow" ?

AdrienHorgnies replied:
Because Registration is more like the whole process. Registration include the licenses the user would want and it's in the other sheet. Registrant is the guy who wants to go through registration, I thought it was more suitable.

rlaloux replied:
And the sheet itself is call "registrants" so, indeed it's fine.

rlaloux commented on src/main/java/com/altissia/clickandrun/domain/spreadsheet/concrete/registration/RegistrantRow.java:
```diff
@@ -0,0 +1,75 @@
+    @Override
+    public int hashCode() {
+        return Objects.hash(getFirstName(), getLastName(), getLogin());
+    }
```
Missing `toString()`

AdrienHorgnies replied:
Done

rlaloux commented on src/main/java/com/altissia/clickandrun/domain/spreadsheet/concrete/registration/RegistrationWB.java:
```diff
@@ -0,0 +1,27 @@
+import com.altissia.clickandrun.domain.spreadsheet.Sheet;
+import com.altissia.clickandrun.domain.spreadsheet.Workbook;
+
+public class RegistrationWB extends Workbook {
```
I know that personal preferences should not be in review but I felt I needed to mention it: i don't like acronym in class name (WB).

AdrienHorgnies replied:
Done

AdrienHorgnies replied:
Done

rlaloux commented on src/main/java/com/altissia/clickandrun/service/extended/WorkbookExtendedService.java:
```diff
@@ -0,0 +1,217 @@
+            if (header == null || header.getPhysicalNumberOfCells() == 0) {
+                sheet.addHeaderError(new FieldValidation("header", "", "com.altissia.constraints.header.empty"));
+            } else {
+                Set<String> headerStringCellValues = StreamSupport.stream(header.spliterator(), false).map(Cell::getStringCellValue).collect(Collectors.toSet());
```
I would create a specific `HeaderValidation` which extends `FieldValidation` and add the column index.

Why? Because in some case, you might found an existing but empty header at a high column index, and to find it by youself as an user based on your validation, it might be tricky.

AdrienHorgnies replied:
Done, I also introduced a bit more of logic to catch more cases such as blank header and duplicate header.

AdrienHorgnies replied:
Done

rlaloux commented on src/test/java/com/altissia/clickandrun/web/rest/extended/ImportGenericValidationTest.java:
```diff
@@ -0,0 +1,184 @@
+
+    @Before
+    public void setup() {
+        RegistrationResource registrationResource = new RegistrationResource(workbookExtendedService);
```
This "generic" unit test uses a specific resource, which is fine, but it should also be a specific unit-test for this resource covering the validation of the particular xlsx and specific constraint related to this xlsx (duplicate login, duplicate cell, ...)

AdrienHorgnies replied:
Tested in next PR "nominal case". My purpose here was only to test general behaviour.

rlaloux commented on src/main/java/com/altissia/clickandrun/domain/spreadsheet/Workbook.java:
```diff
@@ -0,0 +1,83 @@
+	 * @param sheetName the name of the sheet
+	 * @return the model related to the sheet
+	 */
+	public List<? extends Row> getSheetRows(String sheetName) {
```
This method will be subject to uncheckd cast in all processor.
I think you are better off doing the uncheckd cast here instead of doing it in many classes:

```java
	public <T extends Row> List<T> getSheetRows(String sheetName) {
        return (List<T>) this.sheets.get(sheetName).getRows();
	}
```

AdrienHorgnies replied:
Done

rlaloux reviewed (commented):
Enjoy

Commit:
null safe FieldValidation::compareTo

Commit:
registrant row toString

Commit:
WB suffix -> Workbook

Commit:
rebranding

Commit:
introduce HeaderValidation to get column index in error

Commit:
add column getter for jackson

Commit:
handle missing header differently

I found that empty was confusing and it lacked information
I prefer to switch to standard logic for single header missing
Only you'll have one for each item of the header

Commit:
test blank header

Commit:
test real missing header (no XML tag at all)

Commit:
generify getSheetRows to avoid unchecked cast everywhere
